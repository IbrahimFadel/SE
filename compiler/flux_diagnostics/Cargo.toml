[package]
name = "flux_diagnostics"
version = "0.1.0"
edition = "2021"

[dependencies]
ariadne = "0.1.5"
lasso = { version = "0.6.0", features = ["multi-threaded"] }
itertools = "0.10.5"
flux_span = { path = "../flux_span" }