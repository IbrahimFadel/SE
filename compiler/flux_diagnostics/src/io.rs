use colored::Colorize;
use itertools::Itertools;

use crate::DiagnosticCode;

/// IO Errors have no files and therefore we can't use ariadne for them
#[derive(Debug)]
pub struct IOError {
    code: DiagnosticCode,
    msg: String,
    help: Vec<String>,
}

impl IOError {
    pub fn new(code: DiagnosticCode, msg: String, help: Vec<String>) -> Self {
        Self { code, msg, help }
    }

    pub fn to_string(self) -> String {
        format!(
            "{} {}\n\t{}",
            format!("[{}]", self.code).red(),
            self.msg.red().bold(),
            self.help.iter().map(|s| s.blue()).join("\n\t")
        )
    }

    pub fn report(self) {
        eprintln!("{}", self.to_string());
    }
}
