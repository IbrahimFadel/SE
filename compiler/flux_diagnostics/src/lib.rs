mod diagnostic;
pub mod fmt;
mod io;
mod reporting;

use std::fmt::Display;

use colored::Colorize;
pub use diagnostic::*;
pub use io::IOError;
pub use reporting::*;

pub fn ice(msg: impl Display) -> ! {
    panic!(
        "{}\n\t{}",
        "internal compiler error:".red(),
        format!("{}", msg).blue()
    )
}
