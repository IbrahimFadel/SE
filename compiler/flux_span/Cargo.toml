[package]
name = "flux_span"
version = "0.1.0"
edition = "2021"

[dependencies]
text-size = "1.1.1"
cstree = "0.12.0"
lasso = "0.7.2"
