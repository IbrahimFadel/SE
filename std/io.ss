fn print() -> i32 {
    return 0;
}