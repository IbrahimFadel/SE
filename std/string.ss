struct string {
    i8 *buffer;
    i64 length;
    i64 maxLength;
    i64 factor;
};

fn string_create_default(string *this) -> void {
    this->buffer = nullptr;
    this->length = 0;
    this->maxLength = 0;
    this->factor = 16;
}

fn string_delete(string *this) -> void {
    i8 *buf = this->buffer;
    if(buf != nullptr) {
        free(buf);
    }
}

fn string_resize(string *this, i64 value) -> void {
    i8 *output = malloc(value);

    i8 *buf = this->buffer;
    i64 len = this->length;

    memcpy(output, buf, len);

    free(buf);
    this->buffer = output;
}

fn string_add_char(string *this, i8 value) -> void {
    i64 len = this->length;
    i64 maxLen = this->maxLength;

    if(len == maxLen) {
        i64 factor = this->factor;
        string_resize(this, maxLen + factor);
    }

    i8 *buf = this->buffer;
    buf[len] = value;
    i64 newLen = len + 1;
    this->length = newLen;
}