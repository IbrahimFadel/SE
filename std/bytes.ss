import "./io.ss";

fn bytesFunction() -> void {
    
}