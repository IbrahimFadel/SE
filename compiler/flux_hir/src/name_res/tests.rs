use std::io::{BufWriter, Write};

use flux_diagnostics::{reporting::FileCache, Diagnostic};
use flux_span::FileId;
use la_arena::Idx;
use lasso::ThreadedRodeo;
use once_cell::sync::Lazy;
use pretty::BoxAllocator;

use crate::{
    body::LoweredBodies, hir::Function, item_tree::ItemTree, ModuleDefId, ModuleId, TypeInterner,
};

use super::{mod_res::FileResolver, DefMap};

static STRING_INTERNER: Lazy<ThreadedRodeo> = Lazy::new(ThreadedRodeo::new);
static TYPE_INTERNER: Lazy<TypeInterner> = Lazy::new(|| TypeInterner::new(&STRING_INTERNER));

fn check(content: &str) -> (DefMap, LoweredBodies, Vec<Diagnostic>) {
    let mut file_cache = FileCache::new(&STRING_INTERNER);
    let files = content.split("//-").skip(1);
    let mut entry_file_path = None;
    for (i, file) in files.enumerate() {
        let newline_loc = file
            .find("\n")
            .expect("malformated input to `check` function in name resolution unit test");
        let file_path = &file[..newline_loc];
        let content = &file[newline_loc..];
        file_cache.add_file(file_path, content);
        if i == 0 {
            entry_file_path = Some(file_path);
        }
    }
    let (def_map, mut diagnostics) = match entry_file_path {
        Some(entry_path) => crate::build_def_map(
            entry_path,
            &mut file_cache,
            &STRING_INTERNER,
            &TYPE_INTERNER,
            &TestFileResolver,
        ),
        None => panic!("malformated input to `check` function in name resolution unit test"),
    };
    let (lowered_bodies, mut diagnostics2) =
        crate::lower_def_map_bodies(&def_map, &STRING_INTERNER, &TYPE_INTERNER);
    diagnostics.append(&mut diagnostics2);
    (def_map, lowered_bodies, diagnostics)
}

fn fmt_file_id(file_id: FileId, string_interner: &'static ThreadedRodeo) -> &str {
    string_interner.resolve(&file_id.0)
}

fn fmt_function(
    f: &Function,
    f_idx: Idx<Function>,
    module_id: ModuleId,
    allocator: &BoxAllocator,
    buf: &mut BufWriter<Vec<u8>>,
    lowered_bodies: &LoweredBodies,
) {
    f.pretty::<_, ()>(&allocator, &STRING_INTERNER, &TYPE_INTERNER)
        .1
        .render(50, buf)
        .unwrap();
    let body = lowered_bodies
        .indices
        .get(&(module_id, ModuleDefId::FunctionId(f_idx)))
        .unwrap();
    body.pretty::<_, ()>(
        &allocator,
        &STRING_INTERNER,
        &TYPE_INTERNER,
        &lowered_bodies.exprs,
    )
    .1
    .render(50, buf)
    .unwrap();
    buf.write("\n\n".as_bytes()).unwrap();
}

fn fmt_item_tree(
    module_id: ModuleId,
    item_tree: &ItemTree,
    lowered_bodies: &LoweredBodies,
) -> String {
    let allocator = BoxAllocator;
    let mut buf = BufWriter::new(Vec::new());
    item_tree.functions.iter().for_each(|(f_idx, f)| {
        fmt_function(f, f_idx, module_id, &allocator, &mut buf, lowered_bodies);
    });
    let bytes = buf.into_inner().unwrap();
    String::from_utf8(bytes).unwrap()
}

fn fmt_def_map(def_map: &DefMap, lowered_bodies: &LoweredBodies) -> String {
    let mut item_tree_s = String::from("Item Tree\n\n");
    let mut mod_map_s = String::from("Module Id Map\n\n");
    for (module_id, item_tree) in def_map.item_trees.iter() {
        let parent = def_map[module_id].parent;
        if let Some(parent) = parent {
            mod_map_s += &format!("{} -> {}\n", parent.into_raw(), module_id.into_raw());
        }

        let file_id = def_map[module_id].file_id;
        item_tree_s += &format!(
            "File Id: {}\nModule Id: {}\n\n",
            fmt_file_id(file_id, &STRING_INTERNER),
            module_id.into_raw()
        );
        item_tree_s += &fmt_item_tree(module_id, item_tree, lowered_bodies);
    }
    format!("{}\n\n{}", mod_map_s, item_tree_s)
}

fn no_errors(content: &str) {
    let (def_map, lowered_bodies, diagnostics) = check(content);
    assert_eq!(diagnostics.len(), 0);
    let s = fmt_def_map(&def_map, &lowered_bodies);
    insta::assert_snapshot!(s);
}

#[test]
fn basic() {
    no_errors(
        r#"
//- ./main.flx

mod foo;
mod bar;

fn main() -> i32 {
    0
}

//- ./foo.flx

pub fn inside_foo() {}

fn also_inside_foo(x X, y i32) -> u32 {
    0
}

//- ./bar.flx

mod bazz;

fn inside_bar() {}

//- ./bar/bazz.flx

fn inside_bazz() {

}
"#,
    );
}

struct TestFileResolver;

impl FileResolver for TestFileResolver {
    fn resolve_absolute_path(
        &self,
        path: &str,
        file_cache: &mut FileCache,
    ) -> Option<(flux_span::FileId, String)> {
        Some(file_cache.get_by_file_path(path))
    }

    fn resolve_relative_path(
        &self,
        path: super::mod_res::RelativePath,
        file_cache: &mut flux_diagnostics::reporting::FileCache,
    ) -> Option<(flux_span::FileId, String)> {
        let anchor_path = file_cache.get_file_dir(&path.anchor);
        let absolute_path = format!("{anchor_path}/{}", path.path);
        Some(file_cache.get_by_file_path(&absolute_path))
    }
}
