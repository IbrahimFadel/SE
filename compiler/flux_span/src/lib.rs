use std::{borrow::Borrow, hash::Hash, ops::Deref};

use itertools::Itertools;
use lasso::{Spur, ThreadedRodeo};
use text_size::TextRange;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Span {
    pub range: TextRange,
}

impl Span {
    pub fn new<T: Into<std::ops::Range<u32>>>(range: T) -> Self {
        let std::ops::Range { start, end } = range.into();
        let range: TextRange = TextRange::new(start.into(), end.into());
        Self { range }
    }

    /// Convert an iterator of spanned items into a span of all the items
    ///
    /// `[Span]` -> `Span`
    ///
    /// Returns `None` if the iterator has no items, as there can be no span
    pub fn span_iter_of_span(iter: impl IntoIterator<Item = Span>) -> Option<Span> {
        let mut iter = iter.into_iter().peekable();
        let first = iter.peek()?;
        let start = first.range.start();
        let mut end = first.range.end();
        iter.for_each(|t| end = t.range.end());
        let span = Span::new(TextRange::new(start, end));
        Some(span)
    }

    /// Convert an iterator of spanned items into a span of all the items
    ///
    /// `[Spanned<T>]` -> `Span`
    ///
    /// Returns `None` if the iterator has no items, as there can be no span
    pub fn span_iter_of_spanned<T, B>(iter: impl IntoIterator<Item = B>) -> Option<Span>
    where
        B: Borrow<Spanned<T>>,
    {
        let mut iter = iter.into_iter().peekable();
        let first = iter.peek()?;
        let start = first.borrow().span.range.start();
        let mut end = first.borrow().span.range.end();
        iter.for_each(|t| end = t.borrow().span.range.end());
        let span = Span::new(TextRange::new(start, end));
        Some(span)
    }

    pub fn in_file(self, file_id: FileId) -> InFile<Span> {
        InFile::new(self, file_id)
    }
}

#[derive(Debug, Clone)]
pub struct Spanned<T> {
    pub inner: T,
    pub span: Span,
}

impl<T: PartialEq> PartialEq for Spanned<T> {
    fn eq(&self, other: &Self) -> bool {
        self.inner.eq(&other.inner)
    }
}

impl<T: PartialEq> Eq for Spanned<T> {}

impl<T: Hash> Hash for Spanned<T> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.inner.hash(state)
    }
}

impl<T> Spanned<T> {
    pub fn new(inner: T, span: Span) -> Self {
        Self { inner, span }
    }

    /// Convert an iterator of spanned items into a spanned collection of unspanned items
    ///
    /// `[Spanned<T>]` -> `Spanned<[T]>`
    ///
    /// Returns `None` if the iterator has no items, as there can be no span
    pub fn spanned_iter<C: FromIterator<T>>(
        iter: impl IntoIterator<Item = Spanned<T>>,
    ) -> Option<Spanned<C>> {
        let mut iter = iter.into_iter().peekable();
        let first = iter.peek()?;
        let start = first.span.range.start();
        let mut end = first.span.range.end();
        let c = C::from_iter(iter.inspect(|t| end = t.span.range.end()).map(|v| v.inner));
        let span = Span::new(TextRange::new(start, end));
        Some(Spanned::new(c, span))
    }

    /// Convert an iterator of spanned items into a spanned collection of unspanned items that have been mapped with the supplied closure
    ///
    /// `[Spanned<A>]` -> `Spanned<[B]>`
    ///
    /// Returns `None` if the iterator has no items, as there can be no span
    pub fn spanned_iter_with<C, F, B>(
        iter: impl IntoIterator<Item = Spanned<T>>,
        f: F,
    ) -> Option<Spanned<C>>
    where
        C: FromIterator<B>,
        F: Fn(T) -> B,
    {
        let mut iter = iter.into_iter().peekable();
        let first = iter.peek()?;
        let start = first.span.range.start();
        let mut end = first.span.range.end();
        let c = C::from_iter(
            iter.inspect(|t| end = t.span.range.end())
                .map(|v| f(v.inner)),
        );
        let span = Span::new(TextRange::new(start, end));
        Some(Spanned::new(c, span))
    }

    // Gets the [`Span`] of an iterator of [`Spanned`] items
    //
    // Returns `None` if the iterator has no items, as there can be no span
    // pub fn span_iter<C: FromIterator<T>>(
    //     iter: impl IntoIterator<Item = Spanned<T>>,
    // ) -> Option<Span> {
    //     let mut iter = iter.into_iter().peekable();
    //     let first = iter.peek()?;
    //     let start = first.span.range.start();
    //     let mut end = first.span.range.end();
    //     iter.for_each(|t| end = t.span.range.end());
    //     let span = Span::new(TextRange::new(start, end));
    //     Some(span)
    // }

    pub fn in_file(self, file_id: FileId) -> InFile<Spanned<T>> {
        InFile::new(self, file_id)
    }
}

impl<A> Spanned<A> {
    /// Maps the inner value of a [`Spanned`]
    ///
    /// `Spanned<A>` -> `Spanned<B>`
    ///
    pub fn map<F, B>(self, f: F) -> Spanned<B>
    where
        F: FnOnce(A) -> B,
    {
        Spanned::new(f(self.inner), self.span)
    }

    /// Maps the inner value of an [`Spanned`] passing the values to the closure by reference
    ///
    /// `Spanned<A>` -> `Spanned<B>`
    ///
    pub fn map_ref<F, B>(&self, f: F) -> Spanned<B>
    where
        F: FnOnce(&A) -> B,
    {
        Spanned::new(f(&self.inner), self.span)
    }
}

impl<T> Deref for Spanned<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
pub struct FileId(pub Spur);

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InFile<T> {
    pub inner: T,
    pub file_id: FileId,
}

impl<T> InFile<T> {
    pub fn new(inner: T, file_id: FileId) -> Self {
        Self { inner, file_id }
    }

    /// Maps the inner value of an [`InFile`]
    ///
    /// `InFile<A>` -> `InFile<B>`
    pub fn map<F, B>(self, f: F) -> InFile<B>
    where
        F: FnOnce(T) -> B,
    {
        InFile::new(f(self.inner), self.file_id)
    }

    /// Maps the inner value of an [`InFile`] passing the values to the closure by reference
    ///
    /// `InFile<A>` -> `InFile<B>`
    pub fn map_ref<F, B>(&self, f: F) -> InFile<B>
    where
        F: FnOnce(&T) -> B,
    {
        InFile::new(f(&self.inner), self.file_id)
    }
}

impl<T> InFile<Spanned<T>> {
    /// Maps the inner value of an `InFile<Spanned<T>` passing the values to the closure by reference
    ///
    /// `InFile<Spanned<A>>` -> `InFile<Spanned<B>>`
    pub fn map_inner<F, B>(self, f: F) -> InFile<Spanned<B>>
    where
        F: FnOnce(T) -> B,
    {
        InFile::new(self.inner.map(f), self.file_id)
    }

    /// Maps the inner value of an `InFile<Spanned<T>` passing the values to the closure by reference
    ///
    /// `InFile<Spanned<A>>` -> `InFile<Spanned<B>>`
    pub fn map_inner_ref<F, B>(&self, f: F) -> InFile<Spanned<B>>
    where
        F: FnOnce(&T) -> B,
    {
        InFile::new(self.inner.map_ref(|v| f(v)), self.file_id)
    }
}

impl<A: Clone> InFile<A> {}

impl<T> Deref for InFile<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

pub type FileSpanned<T> = InFile<Spanned<T>>;

pub trait WithSpan: Sized {
    fn at(self, span: Span) -> Spanned<Self> {
        Spanned::new(self, span)
    }

    fn in_file(self, file_id: FileId, span: Span) -> InFile<Spanned<Self>> {
        InFile::new(Spanned::new(self, span), file_id)
    }
}

impl WithSpan for String {}
impl<'a> WithSpan for &'a str {}
impl WithSpan for Spur {}
impl WithSpan for usize {}
impl<T> WithSpan for Vec<T> {}
impl<T, F> WithSpan for std::iter::Map<T, F> {}
// impl<I, T> WithSpan for I where I: Iterator<Item = T> {}

pub trait ToSpan: Into<TextRange> {
    fn to_span(self) -> Span {
        Span::new(self.into())
    }
}

impl ToSpan for TextRange {}

pub fn spur_iter_to_spur<'a>(
    spurs: impl Iterator<Item = &'a Spur>,
    interner: &'static ThreadedRodeo,
) -> Spur {
    interner.get_or_intern(spurs.map(|spur| interner.resolve(spur)).join("::"))
}

// #[macro_export]
// macro_rules! can_be_spanned {
//     ($($name:ident)*) => {
//         $( $name );*
//         // $($name),+,
//     };
// }
// macro_rules! can_be_spanned {
//     (
//         $name:ident,
//         $($rest:tt)*
//     ) => {
//         $(impl WithSpan for $name {})*
//         can_be_spanned! {
//             $($rest)*
//         }
//     };
//     () => {}
// }
