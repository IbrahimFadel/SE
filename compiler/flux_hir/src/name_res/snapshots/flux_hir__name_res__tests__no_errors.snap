---
source: compiler/flux_hir/src/name_res/tests.rs
expression: s
---
Module Id Map

0 -> 1
0 -> 2
2 -> 3


Item Tree

File Id:  ./main.flx
Module Id: 0

fn main() -> i32 {
0
}

File Id:  ./foo.flx
Module Id: 1

pub fn inside_foo() -> () {

}

fn also_inside_foo(x X, y i32) -> u32 {
0
}

File Id:  ./bar.flx
Module Id: 2

fn inside_bar() -> () {

}

File Id:  ./bar/bazz.flx
Module Id: 3

fn inside_bazz() -> () {

}


